crates/guest/src/guest.rs

```diff
@@ -60,7 +60,8 @@ where
     O: serde::de::DeserializeOwned + std::fmt::Debug,
 {
     // Call the host function and receive the length of the serialized result.
-    let input_bytes = holochain_serialized_bytes::encode(&input).map_err(|e| wasm_error!(e))?;
+    let mut input_bytes = holochain_serialized_bytes::encode(&input).map_err(|e| wasm_error!(e))?;
```

Comment:
The allocation/deallocate assumes that the length and capacity of the vector match. That is really only guaranteed to be true when `__hc__allocate_1` is used. The mismatches in length/capacity following an `encode` were resulting in a panic when trying to re-assemble the vector.

Reply:
Totally cool with merging this if it fixes the problem, but [shrink_to_fit](https://doc.rust-lang.org/std/vec/struct.Vec.html#method.shrink_to_fit) says it's a best effort and the specific allocator may still leave excess capacity.

Should we instead pass along / capture the capacity so the deallocator can work correctly?

Reply:
Hmm, yes probably something like that but then we have to make the return values a bit more complex because we're packing the pointer+length into an i64 to return them so we'd need a different type to communicate the allocated space.

I wonder if we change the way the allocation is being done so that it doesn't rely on the internal representation of vector... I'm happy reading memory allocated by a vector because it's just a contiguous block but try to deallocate it is fraught with issues -> https://doc.rust-lang.org/std/vec/struct.Vec.html#method.from_raw_parts_in

Reply:
@ThetaSinner could you write a test that shows the problem?

the invariants stated in the docs aren't that extreme, they don't say that length and capacity have to be the same, just that capacity is more than length

vectors were used in the wasmer docs i'm pretty sure, which is why there are here

Reply:
> vectors were used in the wasmer docs i'm pretty sure, which is why there are here

The examples I could find were mostly allocated with data, which I believe won't over allocate. I didn't find much advice about deallocating